enterprise/server/action_cache_server_proxy/action_cache_server_proxy.go

```diff
@@ -36,19 +50,109 @@ func NewActionCacheServerProxy(env environment.Env) (*ActionCacheServerProxy, er
+		sb.WriteString("o")
+	}
+	for _, s := range req.GetInlineOutputFiles() {
+		sb.WriteString(s)
+	}
+	buf := strings.NewReader(sb.String())
+
+	d, err := digest.Compute(buf, req.GetDigestFunction())
+	if err != nil {
+		return nil, err
+	}
+	return digest.NewACResourceName(d, req.GetInstanceName(), req.GetDigestFunction()), nil
+}
+
+func (s *ActionCacheServerProxy) getLocallyCachedActionResult(ctx context.Context, req *repb.GetActionResultRequest) (*repb.ActionResultWithDigest, error) {
+	key, err := s.getACKeyForGetActionResultRequest(req)
+	if err != nil {
+		return nil, err
+	}
+	out := &repb.ActionResultWithDigest{}
+	err = cachetools.ReadProtoFromActionCache(ctx, s.localCache, key, out)
+	if err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+func (s *ActionCacheServerProxy) cacheActionResultLocally(ctx context.Context, req *repb.GetActionResultRequest, resp *repb.ActionResult) error {
+	key, err := s.getACKeyForGetActionResultRequest(req)
+	if err != nil {
+		return err
+	}
+	d, err := digest.ComputeForMessage(resp, req.GetDigestFunction())
+	if err != nil {
+		return err
+	}
+	valueToStore := &repb.ActionResultWithDigest{
+		Digest:       d,
+		ActionResult: resp,
+	}
+	buf, err := valueToStore.MarshalVT()
+	if err != nil {
+		return err
+	}
+
+	return s.localCache.Set(ctx, key.ToProto(), buf)
+}
+
 // Action Cache entries are not content-addressable, so the value pointed to
-// by a given key may change in the backing cache. Thus, always fetch them from
-// the authoritative cache.
+// by a given key may change in the backing cache. Thus, we always send a
+// request to the authoritative cache, but send a hash of the last value we
+// received to avoid transferring data on unmodified actions.
 func (s *ActionCacheServerProxy) GetActionResult(ctx context.Context, req *repb.GetActionResultRequest) (*repb.ActionResult, error) {
+	ctx, err := prefix.AttachUserPrefixToContext(ctx, s.env)
+	if err != nil {
+		return nil, err
+	}
+
+	// First, see if we have a local copy of this ActionResult.
+	var local *repb.ActionResultWithDigest
+	if *cacheActionResults {
+		var err error
+		local, err = s.getLocallyCachedActionResult(ctx, req)
+		if err != nil && !status.IsNotFoundError(err) {
+			return nil, err
+		}
+		if local.GetDigest().GetHash() != "" {
+			// If there's a hash on the local copy, use it and see if remote has it.
+			req.CachedActionResultDigest = local.GetDigest()
+		}
+	}
+
 	resp, err := s.remoteCache.GetActionResult(ctx, req)
 	labels := prometheus.Labels{
-		metrics.StatusLabel:        fmt.Sprintf("%d", gstatus.Code(err)),
-		metrics.CacheHitMissStatus: "miss",
+		metrics.StatusLabel: fmt.Sprintf("%d", gstatus.Code(err)),
+	}
+
+	// The response indicates that our cached value is valid; use it.
+	if *cacheActionResults && req.GetCachedActionResultDigest().GetHash() != "" &&
+		proto.Equal(req.GetCachedActionResultDigest(), resp.GetActionResultDigest()) {
+		// XXX: Should we re-hash the AR here to prevent tampering?
```

Comment:
is the concern here a race, or something else?

Reply:
my concern is, apparently, not trusting my own code.  because we write out the value with a pre-computed digest, we aren't doing any validation when it pops out.  i could instead either:

- compute the hash of the `ActionResult` every time before I send it to the remote server.
- put a pointer in the AC and read the "actual" actionresult from the CAS instead, which would mean the hash would be validated on the way out of the CAS

Reply:
I would just do 2, I think that's what Tyler recommended in his other comment too?

Reply:
yup, done